feat: SerdeVecMap type for serializing ID maps

hiltontj opened the pull request:
This PR introduces a new type `SerdeVecHashMap` that can be used in places where we need a HashMap with the following properties:
1. When serialized, it is serialized as a list of key-value pairs, instead of a map
2. When deserialized, it assumes the serialization format from (1.) and deserializes from a list of key-value pairs to a map
3. Does not allow for duplicate keys on deserialization

This is useful in places where we need to create map types that map from an identifier (integer) to some value, and need to serialize that data. For example: in the WAL when serializing write batches, and in the catalog when serializing the database/table schema.

This PR refactors the code in `influxdb3_wal` and `influxdb3_catalog` to use the new type for maps that use `DbId` and `TableId` as the key. Follow on work can give the same treatment to `ColumnId` based maps once that is fully worked out.

## Explanation

If we have a `HashMap<u32, String>`, `serde_json` will serialize it in the following way:
```json
{"0": "foo", "1": "bar"}
```
i.e., the integer keys are serialized as strings, since JSON doesn't support any other type of key in maps.

`SerdeVecHashMap<u32, String>` will be serialized by `serde_json` in the following way:
```json,
[[0, "foo"], [1, "bar"]]
```
and will deserialize from that vector-based structure back to the map. This allows serialization/deserialization to run directly off of the `HashMap`'s `Iterator`/`FromIterator` implementations.

## The Controversial Part

One thing I also did in this PR was switch the catalog from using a `BTreeMap` for tables to using the new `HashMap` type. This breaks the deterministic ordering of the database schema's `tables` map and therefore wrecks the snapshot tests we were using. I had to comment those parts of their respective tests out, because there isn't an easy way to make the underlying hashmap have a deterministic ordering just in tests that I am aware of.

If we think that using `BTreeMap` in the catalog is okay over a `HashMap`, then I think it would be okay to roll a similar `SerdeVecBTreeMap` type specifically for the catalog. Coincidentally, this may actually be a good use case for [`indexmap`](https://docs.rs/indexmap/latest/indexmap/), since it holds supposedly similar lookup performance characteristics to hashmap, while preserving order and _having faster iteration_ which could be a win for WAL serialization speed. It also accepts different hashing algorithms so could be swapped in with FNV like `HashMap` can.

## Follow-up work

Use the `SerdeVecHashMap` for column data in the WAL following https://github.com/influxdata/influxdb/issues/25461

hiltontj commented on influxdb3_catalog/src/catalog.rs:
```diff
@@ -1104,9 +1057,19 @@ mod tests {
         // Perform a snapshot test to check that the JSON serialized catalog does not change in an
         // undesired way when introducing features etc.
-        assert_json_snapshot!(catalog);
+        // insta::with_settings!({
+        //     sort_maps => true,
+        //     description => "catalog serialized, this snapshot is good for catching changes in \
+        //     catalog serialization"
+        // }, {
+        //     assert_json_snapshot!(catalog);
+        // });
```
This and a couple of other `insta` snapshots are commented out in this PR. We likely still want to be doing some kind of snapshot testing, but as per the `TODO` comment above these lines, this is not possible when using a `HashMap` like this.

hiltontj commented on influxdb3_catalog/src/catalog.rs:
```diff
@@ -466,7 +420,7 @@ pub struct DatabaseSchema {
     pub name: Arc<str>,
     /// The database is a map of tables
-    pub tables: BTreeMap<TableId, TableDefinition>,
+    pub tables: SerdeVecHashMap<TableId, TableDefinition>,
```
Changing this to use a `HashMap` instead of `BTreeMap` for faster lookups.

hiltontj commented on influxdb3_id/src/serialize.rs:
```diff
@@ -0,0 +1,207 @@
+/// During deserialization, there are no duplicate keys allowed. If duplicates are found, an error
+/// will be thrown.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SerdeVecHashMap<K: Eq + std::hash::Hash, V>(HashMap<K, V>);
```
If this is mainly used for ID maps, which are integers, we might be able to use FNV hash: https://doc.servo.org/fnv/

hiltontj replied:
I opened https://github.com/influxdata/influxdb/issues/25494

hiltontj reviewed (commented):
Left some comments in-line.

pauldix reviewed (approved):
Looks good. My thought on the snapshot thing is something like this: we want to test serialization and deserialization. Anything that is going to be a file format needs to be versioned. So I'd think that we'd want:

* A static file with the appropriate data members checked into the repo
* A test that can deserialize the static file and validate it has the members we expect (this is all hard coded stuff)
* Another test that takes an in memory structure, serializes it, then deserializes that and validates they're the same (this might be unnecessary)

hiltontj commented on influxdb3_catalog/src/catalog.rs:
```diff
@@ -1104,9 +1057,19 @@ mod tests {
+        // TODO: with the use of the SerdeVecHashMap, using a snapshot test is flaky. This is
+        // because the serialized vector can not have a deterministic order. So, this part of the
+        // test is disabled for now.
+
```
Yeah, `insta` is compelling but not the  right tool here. I think @waynr was grappling with a similar issue in clustered so I may pick his brain to see if he landed on a solution.

hiltontj replied:
I opened https://github.com/influxdata/influxdb/issues/25493

hiltontj commented:
Given the review comments, I went and cleaned up the commented out code and removed the now unused snapshot files in `influxdb3_catalog` in https://github.com/influxdata/influxdb/pull/25492/commits/06e9689984cd853d802bfb49f06314f8ee8187a0

Once ✅ I will merge.
